File: pysrc/xtallography/unit_cell/rhombohedral.py
#   Copyright 2025 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
"""
The rhombohedral module defines classes and methods specific to rhombohedral lattices.
"""

# --- Imports

# Standard library
import copy
import math
from typing import Optional

# Local packages/modules
from xtallography.symmetry import Centering, LatticeSystem

from .. import _JL
from .lattice_constants import RhombohedralLatticeConstants
from .unit_cell import UnitCell
from .unit_cell_symmetry import UnitCellSymmetry


# --- Classes


class RhombohedralUnitCell(UnitCell):
    """
    Class representing a rhombohedral unit cell
    """

    # ---Initializer

    def __init__(
        self,
        a: float,
        alpha: float,
        centering: Optional[Centering] = Centering.PRIMITIVE,
        symmetry_elements: Optional[set] = copy.deepcopy(set()),
    ):
        """
        Initialize RhombohedralUnitCell object.

        Parameters
        ----------
        `a`, `alpha`: lattice constants

        `centering`: centering of unit cell

        `symmetry_elements`: symmetry elements that unit cell is invariant under
        """
        # --- Check arguments

        if a <= 0:
            raise ValueError(f"`a` must be positive. (a={a})")

        if alpha <= 0 or alpha >= 2 * math.pi:
            raise ValueError(
                f"`alpha` must lie in the interval (0, 2 pi). (alpha={alpha})"
            )

        # --- Initialize parent class

        lattice_constants = RhombohedralLatticeConstants(a, alpha)
        symmetry = UnitCellSymmetry(centering, symmetry_elements)
        super().__init__(LatticeSystem.RHOMBOHEDRAL, lattice_constants, symmetry)

    # --- Properties

    @property
    def a(self):
        """
        Return `a`.
        """
        return self.lattice_constants.a

    @property
    def alpha(self):
        """
        Return `alpha`.
        """
        return self.lattice_constants.alpha

    # --- Methods

    def to_julia(self):
        """
        Convert RhombohedralUnitCell object to a Julia UnitCell object.
        """
        return _JL.RhombohedralUnitCell(
            self.a,
            self.alpha,
            centering=self.centering.to_julia(),
            symmetry_elements=_JL.Vector(
                [element.to_julia() for element in self.symmetry_elements]
            ),
        )

    @classmethod
    def from_julia(cls, unit_cell_jl: _JL.UnitCell):
        """
        Convert a Julia UnitCell object to a RhombohedralUnitCell object.
        """
        # --- Check arguments

        if not _JL.isa(unit_cell_jl, _JL.RhombohedralUnitCell):
            raise ValueError(
                "`unit_cell_jl` must be a Julia `RhombohedralUnitCell` object. "
                f"(unit_cell_jl={unit_cell_jl})."
            )

        # --- Convert unit_cell_jl to a RhombohedralUnitCell object

        unit_cell_symmetry = UnitCellSymmetry.from_julia(unit_cell_jl.symmetry)
        unit_cell = RhombohedralUnitCell(
            unit_cell_jl.lattice_constants.a,
            unit_cell_jl.lattice_constants.α,
            centering=unit_cell_symmetry.centering,
            symmetry_elements=unit_cell_symmetry.symmetry_elements,
        )

        return unit_cell

    def __repr__(self):
        """
        Return string representation of RhombohedralUnitCell.
        """
        return (
            f"RhombohedralUnitCell(a={self.a},alpha={self.alpha},"
            f"centering={self.centering},"
            "symmetry_elements=["
            f"{','.join(sorted([str(item) for item in self.symmetry_elements]))}"
            "])"
        )

File: src/unit_cell/rhombohedral.jl
#   Copyright 2023 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
"""
Rhombohedral unit cell types and functions
"""
# --- Exports

# Types
export RhombohedralUnitCell, RhombohedralUnitCellDelta

# Constants
export RHOMBOHEDRAL_MIN_ANGLE, RHOMBOHEDRAL_MAX_ANGLE

# --- Constants

const RHOMBOHEDRAL_MIN_ANGLE = 0
const RHOMBOHEDRAL_MAX_ANGLE = 2π / 3

# --- Types

# ------ RhombohedralUnitCell

"""
    RhombohedralUnitCell

Lattice constants for a rhombohedral unit cell

Fields
======
* `a`: length of the edge of the unit cell

* `α`: angle between edges of the unit cell

* `symmetry`: unit cell symmetry
"""
const RhombohedralUnitCell = UnitCell{Rhombohedral}

# Outer constructor
"""
    RhombohedralUnitCell(
        a::Real, α::Real;
        centering::Centering=primitive_centering,
        symmetry_elements::Union{Set,Vector,Nothing}=nothing
    )

Construct a RhombohedralUnitCell object from a set of lattice constants.

!!! note

    No constraints are imposed on `centering`. The unit cell does _not_ to be a valid
    Bravais lattice.

Keyword Arguments
=================
- `centering`: centering of unit cell

- `symmetry_elements`: symmetry elements of crystal
"""
function RhombohedralUnitCell(
    a::Real,
    α::Real;
    centering::Centering=primitive_centering,
    symmetry_elements::Union{Set,Vector,Nothing}=nothing,
)
    # --- Check arguments

    if a <= 0
        throw(DomainError(a, "`a` must be positive"))
    end

    if α <= RHOMBOHEDRAL_MIN_ANGLE || α >= RHOMBOHEDRAL_MAX_ANGLE
        throw(DomainError(α, "`α` must satisfy 0 < α < 2π / 3"))
    end

    # symmetry elements
    # TODO

    # --- Construct and return RhombohedralUnitCell object

    return RhombohedralUnitCell(
        (a=a, α=α); centering=centering, symmetry_elements=symmetry_elements
    )
end

# ------ RhombohedralUnitCellDelta

"""
    RhombohedralUnitCellDelta

Lattice constant deltas for a rhombohedral unit cell

Fields
======
* `Δa`: delta of the length of the edge of the unit cell

* `Δα`: delta of angle between edges of the unit cell
"""
const RhombohedralUnitCellDelta = UnitCellDelta{Rhombohedral}

# Outer constructors
"""
    RhombohedralUnitCellDelta(Δa::Real, Δα::Real)

Construct a RhombohedralUnitCellDelta object from a set of lattice constant deltas.
"""
function RhombohedralUnitCellDelta(Δa::Real, Δα::Real)
    Δlattice_constants = (Δa=Δa, Δα=Δα)
    return RhombohedralUnitCellDelta(Δlattice_constants)
end

# --- Functions/Methods

# ------ UnitCell methods

function basis(unit_cell::RhombohedralUnitCell)
    # Get lattice constants
    lattice_constants_ = lattice_constants(unit_cell)
    a = lattice_constants_.a
    α = lattice_constants_.α

    # Construct basis
    basis_a = Vector{Float64}([a, 0, 0])
    basis_b = Vector{Float64}([a * cos(α), a * sin(α), 0])
    basis_c = Vector{Float64}([
        a * cos(α),
        a * cot(α) * (1 - cos(α)),
        a / cos(0.5 * α) * sqrt(sin(1.5 * α) * sin(0.5 * α)),
    ])

    return basis_a, basis_b, basis_c
end

function volume(unit_cell::RhombohedralUnitCell)
    # Get lattice constants
    lattice_constants_ = lattice_constants(unit_cell)
    a = lattice_constants_.a
    α = lattice_constants_.α

    # Compute volume
    return 2 * a^3 * sin(0.5 * α) * sqrt(sin(1.5 * α) * sin(0.5 * α))
end

function surface_area(unit_cell::RhombohedralUnitCell)
    lattice_constants_ = lattice_constants(unit_cell)
    return 6 * lattice_constants_.a^2 * sin(lattice_constants_.α)
end

function conventional_cell(::Rhombohedral, unit_cell::UnitCell)
    # --- Check arguments

    conventional_cell_check_args(unit_cell)

    # --- Preparations

    # Get lattice constants
    lattice_constants_ = lattice_constants(unit_cell)
    a = lattice_constants_.a
    α = lattice_constants_.α

    # --- Compute IUCr conventional cell

    # Check limiting cases
    if α ≈ π / 3
        # cubic, face-centering, edge length `a` / sin(π/4)
        @debug "hR --> cF"
        return CubicUnitCell(a / SIN_PI_OVER_FOUR; centering=face_centering)

    elseif α ≈ π / 2
        # cubic, primitive-centering, edge length `a`
        @debug "hR --> cP"
        return CubicUnitCell(a; centering=primitive_centering)

    elseif α ≈ ACOS_MINUS_ONE_THIRD
        # cubic, body-centering, edge length `a` / sin(π/3)
        @debug "hR --> cI"
        return CubicUnitCell(a / SIN_PI_OVER_THREE; centering=body_centering)
    end

    # Not a limiting case, so return a copy of the original unit cell
    return UnitCell(unit_cell)
end
